lib/valkyrie/storage/versioned_shrine.rb

```
@@ -0,0 +1,197 @@
+             end
+
+        raise Valkyrie::StorageAdapter::FileNotFound unless shrine.exists?(shrine_id_for(id)) && !id.to_s.include?(VersionId::DELETION_MARKER)
+        Valkyrie::StorageAdapter::StreamFile.new(id: Valkyrie::ID.new(id.to_s.split(VersionId::VERSION_PREFIX).first),
+                                                 io: DelayedDownload.new(shrine, shrine_id_for(id)),
+                                                 version_id: id)
+      end
+
+      # @return VersionId A VersionId value that's resolved a current reference,
+      #   so we can access the `version_id` and current reference.
+      def version_id(id)
+        version_id = VersionId.new(id)
+        return version_id unless version_id.versioned? && version_id.reference?
+        id = Valkyrie::ID.new(id.to_s.split(VersionId::VERSION_PREFIX).first)
+        version_files(id: id).map { |f| VersionId.new(f) }
+                             .reject { |f| VersionId.new(f).deletion_marker? }
+                             .first
+      end
+
+      # Convert a non-versioned file to a version file basing on its last_modified time.
+      # @param id [Valkyrie::ID]
+      def to_version_file(id:)
+        shrine_id = shrine_id_for(id)
+        last_modified = shrine.object(shrine_id).last_modified
+        version_id = VersionId.new(id).generate_version(timestamp: last_modified).id
+        source_object = Aws::S3::Object.new(shrine.bucket.name, shrine_id, client: shrine.client)
+        source_object.move_to("#{shrine.bucket.name}/#{shrine_id_for(version_id)}")
+      end
+
+      # A class that holds a version id and methods for knowing things about it.
+      # Examples of version ids in this adapter:
+      #   * shrine://[resource_id]/[uuid]_v-current
+      #   * shrine://[resource_id]/[uuid]_v-1694195675462560794
+      #   * shrine://[resource_id]/[uuid]_v-1694195675462560794-deletionmarker
+      class VersionId
+        VERSION_PREFIX = "_v-"
```

I would go with something really obvious here like `/versions/` — easy to read and also makes it easy to understand what you're looking at if, for instance, you're looking at MinIO back-end storage, or you dump the objects to a filesystem. It's also similar to how the Fedora backend uses [`/fcr:versions`](https://github.com/samvera/valkyrie/blob/main/lib/valkyrie/storage/fedora.rb#L174).

---

Also, we might want to call it `VERSION_DELIMITER` rather than `VERSION_PREFIX`, since it comes between the UUID and the timestamp.

---

I think it might be good to follow the VersionedDisk pattern with v- as internal versions management strategic so that it's more consistent over Valkyrie.

I change it from using / to _ because In Minio, my practice showing that there is a bug like issue that preventing the version files to be retrieved with its identifier [resource_id]/[uuid] if the original file with object id [resource_id]/[uuid] exists along with a version file like [resource_id]/[uuid]/[version_id]. And in this case, it won't list those versioned files form the Minio UI ([http://localhost:9001](https://urldefense.com/v3/__http:/localhost:9001__;!!Mih3wA!Hw3D5IllkQQpsQfTYvG7e2IPM6DxkEvVc_OqFryIVoR9PpwbrEJBpGAwkLs2Ks3zACbhws_5VFLZZvieI0Tlmw$)). So it might be good to avoid using anything starting with /.

If we want to learn more about how versions are managed in Fedora, the api call [/fcr:versions](https://urldefense.com/v3/__https:/github.com/samvera/valkyrie/blob/main/lib/valkyrie/storage/fedora.rb*L174__;Iw!!Mih3wA!Hw3D5IllkQQpsQfTYvG7e2IPM6DxkEvVc_OqFryIVoR9PpwbrEJBpGAwkLs2Ks3zACbhws_5VFLZZvhOlXS27A$) will return a list of versions in [Memento](https://urldefense.com/v3/__https:/www.mementoweb.org/guide/quick-intro/__;!!Mih3wA!Hw3D5IllkQQpsQfTYvG7e2IPM6DxkEvVc_OqFryIVoR9PpwbrEJBpGAwkLs2Ks3zACbhws_5VFLZZvgHWV2JJw$) time format (with a timestamp pattern yyyyMMddHHmmss). But Internally, Fedora manages object and its versions with [OCFL specification](https://urldefense.com/v3/__https:/wiki.lyrasis.org/display/FEDORA6x/Fedora*OCFL*Object*Structure__;Kysr!!Mih3wA!Hw3D5IllkQQpsQfTYvG7e2IPM6DxkEvVc_OqFryIVoR9PpwbrEJBpGAwkLs2Ks3zACbhws_5VFLZZvid8XiEjA$), which is more complex.

Let's also take into account with the timestamp pattern together.

---

It makes sense that MinIO would have trouble mixing `[resource_id]/[uuid]` and `[resource_id]/[uuid]/[version_id]`, because the underlying structure is a filesystem, and `[resource_id]/[uuid]` can't be both a file and a directory. But, that shouldn't happen if we're migrating the unversioned file first (which is another argument for migrating first, actually).

What I like about `[resource_id]/[uuid]/versions/yyyyMMddHHmmss` is that it works directly as a semantically meaningful REST-style resource route, like `[resource_id]/[uuid]` and `[resource_id]/[uuid]/versions` both still mean something and it's clear what they mean. (Actually something like `[resource_id]/files/[original-filename]` would be even better, but that would break compatibility with `IdPathGenerator`.)

I think it's good to follow the architecture / dataflow of `VersionedDisk` where we can, because it's closer to what we're doing than the versioning support in the Fedora backend, but I don't think there's a good reason to follow the filname / path format. We know we don't want to use the full format (things like `/fa/ke/re/fakeresourceid/v-1745275964937-example.tif`) because the infixed version number would mean we can't easily retrieve all versions of a file with an S3 prefix lookup. We _could_ use the pairtree directory structure, and it might even be helpful with MinIO in principle, but with the hardware SDSC is using [I don't think it matters](https://min.io/docs/minio/linux/administration/object-management.html#object-organization-and-planning), plus, again, it's incompatible with `IdPathGenerator`. As long as we know our paths are going to be pretty different from the `VersionedDisk` paths, I don't see the value in using `v-`.

---

Do you mean we are going to use pattern `[resource_id]/[uuid]/versions/yyyyMMddHHmmss`? I think it's good if we are going to expose it to the external world. In Fedora, it follows the Memento time gate specification for outer exchange. But for internal implementation, it's OCFL specification.

Maybe if we think more about the difference of outer exchange and internal implementation, the answer could be different for the `VersionedDisk` pattern. Since we all know `v-` means a version, I don't see any benefits with /versions/ as delimiter, except that I need to click one more time to the `versions` folder to view my files in MinIO. And I think keeping it consistent across Valkyrie could be a plus. For external exchanges or Ui display, we can always translate it to a format that make much senses for human communication. 

If we won't care about the case of missing versions issue I describe earlier, I think using slash and follow the `VersionedDisk` pattern to prefix `v-` looks good to me too. That is, something like `/v-` as version delimiter and prefix. In MinIO, the file structure will looks like:
```
/5/47e88f4a-6230-4607-a144-fbfc631b380f/
v-1743388333965
v-1743388333000
v-1743388334000
```
If using /versions/ as delimiter and we follow the current `VersionedDisk` timestamp pattern, it'll looks like

```
/5/47e88f4a-6230-4607-a144-fbfc631b380f/versions/
1743388333965
1743388333000
1743388334000
```

But I think either way will work unless someone strongly agains it.

---

> For external exchanges or Ui display, we can always translate it to a format that make much senses for human communication.

True, but in the long term, I think the easier it is to understand the data even without the software, the better. If a user reports that something happened to their files, and I can just look directly at the storage back-end and see where the files are, or where they should be if they're missing, that's better than if I have to run some code to tell me that, especially if the code might have changed or if that code might be where the issue is.

That said, given that with `[resource_id]/[uuid]` we're already not using REST-type paths, I suppose it doesn't really matter what the prefix is, and it doesn't have to be a directory.

I would still prefer a "Memento plus milliseconds" timestamp format, though, so if someone says they uploaded the file on Thursday March 5 2026, I can just look for "`20260305`", and not have to first do some calculations and then look for "numbers between `1772668800000` and `1772755199999`".

---

I think I am fine with either way. I just think more about the consistency issue and the Valkyrie product as a whole so that we don't need to explain it to users why a version number looks differently. And if we are talking about the time that a file uploaded, it looks like it won't be an issue since every file should have the last modified date there.

With that said, maybe we can have a ticket for a global versioning label format change to follow up? I am mot sure. But I can just change it to the format you want.

---

> I would still prefer a "Memento plus milliseconds" timestamp format, though, so if someone says they uploaded the file on Thursday March 5 2026, I can just look for "`20260305`", and not have to first do some calculations and then look for "numbers between `1772668800000` and `1772755199999`".

This is a convincing argument, love it.

---

Okay. I will update it to use "Memento plus milliseconds" timestamp format: `%Y%m%d%H%M%S%L`. What's the path pattern we want, `[resource_id]/[uuid]/versions/[timestamp]` or a shorter path with `v-` like `[resource_id]/[uuid]/v-[timestamp]`?

---

I think if we're going to use `/` at all (which I think is OK, even with the MinIO issues, because we should be migrating files anyway, and if we're not, it's probably good to fail fast), then we might as well make it a directory-either `/versions/` or, if we want something short, just `/v/`. Shorter might be better.

---

Yes, shorter `/v/` does look better than `/versions/` to me. 

It looks like the following questions are still open:
- Do we want to follow `VersionedDisk` for version label pattern like `v-[timestamp]`?
- Do want another nested folder like `v` to organize the version files for a file/FileMetadata or not? In MinIO, we need to click one more time on the nested `v` folder to look up our file(s).

---

I guess the more I think about it, the more `[resource_id]/[uuid]/v-[timestamp]` starts to make sense - we don't want to make it `[uuid]_v-[timestamp]` (without `/`) because it makes it too hard to look at versions for just a single file (that is, a single UUID); and if we're not saying `[resource_id]/files/[uuid]`, then having a semantic `/v/` directory doesn't add much, and it does mean one more click.

---

That sounds good to go with `/v-`. Since we are converting the existing non-versioned files with the base identifier to version files now, I think we just need to add a note somewhere to document that specific MinIO issue for versions lookup with `/` in the path, which could fail to list other versions if that happens somehow.